File: crates/taflow-core/src/indicators/cumulative_count.rs
use crate::stream::StreamingIndicator;

/// Persistent one-based count of processed observations.
#[derive(Debug, Clone, Default)]
pub struct CumulativeCount {
    count: usize,
    value: Option<f64>,
}

impl CumulativeCount {
    /// Create an empty counter.
    pub fn new() -> Self {
        Self::default()
    }

    /// Count one aligned observation and return the new one-based count.
    pub fn append(&mut self, _input: f64) -> f64 {
        self.count += 1;
        let value = self.count as f64;
        self.value = Some(value);
        value
    }

    /// Extends the state with an aligned slice.
    pub fn extend_slice_into(&mut self, input: &[f64], output: &mut Vec<f64>) {
        output.reserve(input.len());
        output.extend(input.iter().map(|&value| self.append(value)));
    }

    /// Return the latest cumulative count.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Reset the count to its empty state.
    pub fn reset(&mut self) {
        self.count = 0;
        self.value = None;
    }
}

impl StreamingIndicator for CumulativeCount {
    type Output = f64;

    fn append(&mut self, input: f64) -> Option<f64> {
        Some(Self::append(self, input))
    }

    fn value(&self) -> Option<f64> {
        Self::value(self)
    }

    fn reset(&mut self) {
        Self::reset(self);
    }

    fn extend_slice_into(&mut self, input: &[f64], output: &mut Vec<f64>) {
        Self::extend_slice_into(self, input, output);
    }
}

File: python/taflow/indicators/cumulative_count.py
"""Native-backed cumulative observation count adapter."""

from typing import Any

import numpy as np

from .._native import CumulativeCount as _NativeCumulativeCount
from .._series import as_float64_series


class CumulativeCount:
    """Emit the one-based count of chronological observations.

    ``_input`` is the required input series and may be empty to create a fresh
    stream. Each bar emits its one-based position, so ``compute`` returns a
    one-dimensional float NumPy array and ``value`` is ``None`` before the
    first append. ``append``, ``extend``, and ``reset`` mutate and return this
    adapter; the native Rust state performs all counting and warm-up handling.
    """

    def __init__(self) -> None:
        self._state = _NativeCumulativeCount()

    def append(self, _input: float) -> "CumulativeCount":
        """Append one chronological observation and return this adapter."""
        self._state.append(float(_input))
        return self

    def extend(self, _input: Any) -> "CumulativeCount":
        """Append a converted chronological observation history."""
        self._state.extend(as_float64_series(_input))
        return self

    def compute(self) -> np.ndarray:
        """Return the aligned one-based observation counts."""
        return self._state.compute()

    @property
    def value(self) -> float | None:
        """Return the latest count, or ``None`` before the first append."""
        return self._state.value

    def reset(self) -> "CumulativeCount":
        """Restore fresh native state and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of processed observations."""
        return len(self._state)


__all__ = ["CumulativeCount"]
